### ISSUE TITLE
Improve Security Awareness

### ISSUE THREAD

**Issue by WebReflection** [alpinejs/alpine#237]
```js
for (const el of document.querySelectorAll('*')) {
  const attributes = [...el.attributes].filter(attr => '@click' === attr.name);
  if (attributes.length)
    attributes[0].value = 'alert("seriously? in 2020?")';
}
```

You can test it [in codepen](https://codepen.io/WebReflection/pen/JjdrwJZ?editors=1000), and the attack regards any script that might be injected in a page that enabled evaluation for _alpine_.

I think the architecture behind _alpine_ relies entirely on evaluation and deprecated ECMAScript features, and I am a libraries author myself, but I think it's about the time **security** becomes a major Web concern.

I am not sure how to fix this, as `with(...)` statement and evaluation of only the right thing would blow the size of this library, as it would require a proper AST crawling per each attribute content, but I hope this issue would be at least documented properly on top of the README page, as it's way too easy to inject arbitrary code when this library is trusted, and around.

Thanks for making the Web a better place.

**Comment by SimoTod** [comment]
@WebReflection 

I may be missing something but what the difference between your example and
```html
<div>
    <button>Open Dropdown (whatever lib we use)</button>
</div>
<script> <!-- type="evil" -->
for (const el of document.querySelectorAll('button')) {
  el.addEventListener("click", function() {
    alert("seriously? in 2020?")
  })
}
</script>```

The attack you describe assumes you can inject random code in the page. 
If you have that power nobody will stop you from doing more sophisticated things, right?
Shouldn't the exploit describe how you get to that point? Maybe using input with model binding on stuff like that.

**Comment by WebReflection** [comment]
Another example, `ads.js` knows that the site is tracking links to ads and the reactive nature of _alpine_ is in charge of updating stats behind the scene ... `ads.js` searches for special `@click` attributes and changes their behavior ... does the host know about what `ads.js` does? not at all, it's in charge of showing ads.

Same goes with shopping charts, a malicious script wants you to fail only of you click a specific item, and the whole logic is hooked within _alpine_ ... `selectItem = false` as attribute value, and nothing would ever happen.

In few words, in a Web world where "_first come, first serve_", in terms of listeners, _alpine_ is enabling any library to make the site look like it's acting normally, hence unobtrusive, while potentially taking over the whole logic.

These are just few of the reasons the Web doesn't like `eval` and `Function`, where arguably the latter one is safer, as it doesn't expose the whole internal scope, and yet there are reasons evaluation, even via `Function` are not allowed by default in CSP.

Would it be too much to ask to at least specify in your README that **[CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)** must have special care because otherwise nothing works in here?

There's not a single entry about security and _CSP_ needs, which is not how the Web should be in these days, imho.

**Comment by SimoTod** [comment]
Thanks for your detailed explanation @WebReflection. 
I replied to know your point of view since I was interested (I asked another guy im the past but he never got back to me). 
I'm not the owner of the library but I'm sure it will be considered by Caleb. 

Since it's an open source peojwct, you can peobably PR the disclaimer yourself if you like. 

Random considerations... 

I'm not a fan of the x-html directive myself and I agree that it shouldn't even be there.

Alpine requires unsafe-eval so obviously I wouldn't use it to build a banking solution, it will never pass a pentest. I believe its natural use case is for personal blogs and similar where people don't want to go mad with the virtual DOM only to have a dropdown or a fancier modal window.

Thanks again.

**Comment by WebReflection** [comment]
FWIW, I've never used, nor needed, virtual DOM on the Web, and I understand the easy entry point use case for _alipne_, however this might get just more popular, so that some awareness regarding its constrains, or architecture model, *might* help people make a more conscious decision, so I hope Caleb will think about it.

**Comment by nyariv** [comment]
I agree with @WebReflection, alpine js bypasses any protection you try to implement with CSP. The whole idea of CSP is to give blanket protection _because_ it is near impossible to plug all current and future vulnerabilities in your tech stack to prevent any kind of html injection.

Unfortunately using `with()` or proxies is not going to help much and are pretty trivial to bypass if the expression goes through `eval` or `Function`. The only way to really protect yourself is using a custom evaluation function, or wait for [Realms](https://github.com/tc39/proposal-realms) to be adopted by ECMAScript.

I too was inspired by the ideas behind alpine js and started creating my own js eval function ([sandboxjs](https://github.com/nyariv/SandboxJS)) to begin solving that problem. It currently works only on single expressions without code blocks, but allows blacklisting globals or anything in the prototype chain. Unfortunately any kind of solution like this would bloat alpine js considerably which goes against its premise.

**Comment by WebReflection** [comment]
worth saying, that for what I could tell, expressions are always like `thingy = value`, which is a very simple thing to `.split(/\s*=\s*/)`, to take `thingy` as property to use to assign `value` to data, and value is eventually the only thing that might need parsing, if not accepted as JSON right away.

All I am saying is that likely there's no need for a whole sandbox / parser here, but I also don't know this library enough.

**Comment by nyariv** [comment]
its very easy to obfuscate js, so regex checking is not going to help. You could, for example, put `=` sign in a string.

**Comment by WebReflection** [comment]
true that, although if code gets obfuscated here the logic break, as the `with($data)` statement requires exact names ... the split can also be `$data[slit[0].trim()] = split.slice(1).join('=');` but again, I don't think this works, 'cause this library really ask for evaluated code in attributes.

**Comment by trusktr** [comment]
@WebReflection Do any of these concerns also apply to the similar Mavo library? https://mavo.io

**Comment by WebReflection** [comment]
@trusktr I don't want to speculate on @LeaVerou here for at least two reasons:

  * if Lea's code suffers same CSP issues, there should be a bug in Lea's repository, not a discussion in here
  * as much as I've heard about _mavo_, I haven't tried it, read it source, or anything, so I have no idea if it has any issue whatsoever

That being said, since you pointed me at Lea's work, it looks like _mavo_ uses a specialized syntax that has nothing to do with JS, examples:

```
[count(done)] done out of [count(task)] total
[readable(to(filename(image), '.'))]
```

there are not eval-able expressions, as I believe _mavo_ knew better than just polluting the global scope with functions such as `to`, `readable`, etcetera.

However, if _mavo_ is based on `eval` or `Function`, and its eval'd data comes from the shared document, _mavo_ might be as well a vector to inject arbitrary code.

As reminder, [CSP might allow specific libraries and their code to be trusted](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#strict-dynamic_2), if these expose to third part scripts a way to use these libraries via regular, and always allowed, DOM manipulation, then any of these libraries should put a warning on top of their readme, about the risk such library enables in any public site using it, as we, Web developers, don't get to control what other scripts could land in a page, but this is the far-west when it comes to ads injectors (I just happen to work in the field that tries to circumvent these scripts).

**Comment by nyariv** [comment]
Mavo does use `Function` using `with()` https://github.com/mavoweb/mavo/blob/bfe96875e58f8954c7ca2b487df96b6abde8df65/src/mavoscript.js#L792

**Comment by WebReflection** [comment]
@nyariv at least somebody has been honest in there:
https://github.com/mavoweb/mavo/blob/bfe96875e58f8954c7ca2b487df96b6abde8df65/src/mavoscript.js#L782-L785

> // Yes this is a horrible, horrible hack and I’m truly ashamed.
> // If you understand the reasons and can think of a better way, be my guest!

here there's not even a comment about being a bad hack

**Comment by tlgreg** [comment]
A notice that the library relies on `unsafe-eval` would indeed be good in the readme.

The evaluation code came from Vue 2, which solves this issue by having the option to pre-compile the template and only ship the runtime to the client.

Vue 1.0 (when there was no pre-compilation yet) solved this issue by using `notevil` to evaluate expressions on a separate CSP build.

**Comment by bep** [comment]
I would assume that it would be possible in a future version so add a "safe option" that used a simpler eval, which would allow:

```html
<div x-data="{ tab: 'foo' }">
    <button :class="{ 'active': tab === 'foo' }" @click="tab = 'foo'">Foo</button>
    <button :class="{ 'active': tab === 'bar' }" @click="tab = 'bar'">Bar</button>

    <div x-show="tab === 'foo'">Tab Foo</div>
    <div x-show="tab === 'bar'">Tab Bar</div>
</div>
```

But would error on something like:

```html
<div x-data="{ open: false }">
    <button
        @mouseenter.once="
            fetch('/dropdown-partial.html')
                .then(response => response.text())
                .then(html => { $refs.dropdown.innerHTML = html })
        "
        @click="open = true"
    >Show Dropdown</button>

    <div x-ref="dropdown" x-show="open" @click.away="open = false">
        Loading Spinner...
    </div>
</div>
```

**Comment by WebReflection** [comment]
@bep yes, if expressions are simple enough to be parse-able, the former wouldn't need any eval/Function, but the later one would require a hell of a parser

**Comment by tlgreg** [comment]
Was just thinking if `notevil` could be a solution, but probably only similarly to vue1, as a *csp build*.

`notevil` is more than double the size of `alpinejs` itself because it includes `esprima` to parse the javascript expression (`alpinejs` 6.6kB vs `notevil` 15.4kB min+gzip).

Other downsides:
- Performance overhead.
- esprima only supports ECMAScript 2016, so syntax limitations in inline expressions.

**Comment by SimoTod** [comment]
Acorn could be a better alternative (it should be smaller and faster) but
we would need to write our own parser. I assume we could impose some
limitations, too. Something like: Only x-data can access the global scope,
directives should be a single expression (assignment to a variable or call
to a function defined in x-dara), etc.
That would probably "annoy" a lot of people though.

I believe that, if the parser replicates what eval does, even if we have
"fixed" the csp settings, it will still be suffer from the same security
issue.

**Comment by WebReflection** [comment]
@SimoTod yes, a parser that ends up evaluating would have exact same result, and so would globally defined callbacks.

One way to go would be to register once callbacks, i.e.

```js
alpine.register("drop-partial-fetch", $refs => {
  fetch('/dropdown-partial.html')
    .then(response => response.text())
    .then(html => { $refs.dropdown.innerHTML = html })
});
```
After that, all one should do is to write:
```html
<div x-data="{ open: false }">
    <button
        @mouseenter.once="drop-partial-fetch"
        @click="open = true"
    >Show Dropdown</button>

    <div x-ref="dropdown" x-show="open" @click.away="open = false">
        Loading Spinner...
    </div>
</div>
```
That would at least grant that nobody can inject non-registered callbacks or, more important, change their action.

Callbacks are kept private, and no CSP exception is needed. Other attributes with `open = true` can be easily parsed as `key = value` expression without needing eval.

Anyone can still attach arbitrary listeners, eventually, but at least the library wouldn't be the evaluation channel for any kind of good or malicious script.

**Comment by bep** [comment]
If I was to think a little simpler and for the sake of discussion argued that I would need to have a CSP config like this:

```
Content-Security-Policy = "script-src myhost.com 'unsafe-eval'"
```

The above will allow AlpineJS to work, but would disallow constructs like this:

```bash
<script>
for (const el of document.querySelectorAll('*')) {
  const attributes = [...el.attributes].filter(attr => '@click' === attr.name);
  if (attributes.length)
    attributes[0].value = 'alert("seriously? in 2020?")';
}
</script>
```

But then, if you're receiving user-provided content (discussion forum, a "dev blog", whatever) you're worried about people injecting `x-data` components into your DOM ...

What if Alpine added an `Alpine.Init(dataElementFilter)` that restricted Alpine components to live in some defined areas of the page?

You would not solve the CSP problem, but you would strongly reduce the attack surface while still easily support the core uses for Alpine in these situations.

**Comment by SimoTod** [comment]
@bep 
Yeah, in my opinion eval is not necessary evil by itself but requires discipline to stop other scripts from exploiting it (I might be wrong and there could be sophisticated attacks that i have't considered, which is the reason why I'm not comfortable with writing the disclaimer by myself).

I think it's the original point from @WebReflection, if the library needs it, it should, as a first step, just say something in the readme to "educate" and make users aware of it.

Your example is correct but if you have a lot of code, you probably want to put everything in an inline script, like
```javascript
<script>
    function data() {
        return {
            ...
        }
    }
<script>
```

At that point, people without much CSP experience could be tempted to add `unsafe-inline` to the policy where they should really generate a nonce token server side for each request and use that one to trust their safe inline scripts.

Or your policy needs an external library and, if possible, it would be wiser to generate the sha254 of the script and trust that exact implementation rather than whitelisting the external domain and trusting everything that comes from it.

**Comment by bep** [comment]
@SimoTod I think it's fairly easy to avoid `unsafe-inline` with Alpine, and unlike other JS frameworks out there, Alpine does encourage "writing the code yourself" rather than pulling in some random plugin, so that itself limits the attack surface. A way to limit the scope of what DOM nodes can have `x-data`  would limit it even further.

**Comment by iquito** [comment]
Just wanted to throw in that as a new user to Alpine I would have appreciated a heads-up that Alpine relies on eval, as I use a very strict Content-Security-Policy for all my websites, so I cannot use Alpine, but I only found that out after reading through a lot of documentation and finding this issue.

Even better would be a way to use Alpine without eval, as I feel a strict Content-Security-Policy is one of the best ways of making a website secure "by default", so it is a pity when libraries can only be used by turning off these reasonable security features.

**Comment by ryangjchandler** [comment]
> Just wanted to throw in that as a new user to Alpine I would have appreciated a heads-up that Alpine relies on eval, as I use a very strict Content-Security-Policy for all my websites, so I cannot use Alpine, but I only found that out after reading through a lot of documentation and finding this issue.
> 
> 
> 
> Even better would be a way to use Alpine without eval, as I feel a strict Content-Security-Policy is one of the best ways of making a website secure "by default", so it is a pity when libraries can only be used by turning off these reasonable security features.

Well strictly speaking it doesn't use eval. It uses a safer eval, that runs through a constructed function.

**Comment by SimoTod** [comment]
I'm slowly writing a JS sandboxed interpreter which would work with Alpine and remove the need of unsafe-eval.
It might not make it to the core but, in that case, I'm happy to maintain a fork or a plugin for those who need to use Alpine with a strict CSP if there's enough interest in it.
It's going to be based on Acorn (inspired by this article -> https://blog.bitsrc.io/build-a-js-interpreter-in-javascript-using-acorn-as-a-parser-5487bb53390c) and it would add about 30KB (gzipped) to the library.
Initially it will only support a subset of all the possible JS instructions but it should be usable.
I'll try to put everything on Github soon (Hopefully this weekend) in case anyone wants to collaborate.
I'll post an update once ready.

**Comment by nyariv** [comment]
When writing the parser, some considerations that should be taken:

1. prototype pollution: `[].constructor.prototype.filter = () => alert("!")`
2. infinite loops: `while(new Date().getTime() > 0);`
3. global variables access: `window['alert']("!")`
4. slow regex
5. dom injection `el.addEventListener('click', (e) => e.target.closest('body').innerHTML = '!')`

These are just a few of the issues. But like i mentioned in my initial comment here, i have worked on a library that solves these by protecting access on the prototype level and completely forbidding code blocks and loops. I can help out on the parser you are writing and borrow some concepts from [sandboxjs](https://github.com/nyariv/SandboxJS)

**Comment by SimoTod** [comment]
Sure, i had totally forgotten about your initial message. We can maybe use sandboxjs directly. I'll have a look at the code in the next few days.

**Comment by SimoTod** [comment]
@nyariv I think you need to recompile the dist folder.
Your package js defines  `"main": "dist/Sandbox.mjs",` as an entry point but it's the only file that hasn't been updated for 1 month and doesn't match the latest documentation so when I import the Sandbox object, I get an old version.

**Comment by nyariv** [comment]
@SimoTod you are right, it is fixed now.

**Comment by HugoDF** [comment]
@SimoTod @calebporzio since we merged that disclaimer about CSPs should we close this until we want to create a CSP-friendly build.

**Comment by SimoTod** [comment]
I assume we can close it and people can open a new issue if they have further concerns. 

For those who follow, v3 will have a more modular approach which should allow for a  csp friendly build. Unfortunately, this is not possible in v2 without introducing breaking changes.

The readme has now a section about Alpine requiring unsafe eval when using CSP and other security considerations.
It's open to PRs if anybody feels something important is missing.

**Comment by matthewp** [comment]
@SimoTod @HugoDF Now that v3 is out, I haven't been unable to find any documentation on this issue. Is it documented any where?

**Comment by HugoDF** [comment]
> @SimoTod @HugoDF Now that v3 is out, I haven't been unable to find any documentation on this issue. Is it documented any where? 

There's a CSP compatible build in v3 using `Alpine.data`

**Comment by matthewp** [comment]
Good to hear, is this in the docs some where?

**Comment by matthewp** [comment]
Found it! https://alpinejs.dev/advanced/csp

**Comment by matthewp** [comment]
You might be aware, but `@alpinejs/csp` hasn't been published yet.

**Comment by WebReflection** [comment]
FWIW, as initial "*blamer*" of the issue, I am super happy about the current increased awareness and solution: it's elegant, easy to work around, and much cleaner than the default behavior.

Thanks for moving, thinking, and implementing, this forward 👍 ... but if it's still not published, as @matthewp suggested, please do so as soon as you can!

**Comment by jtallinger** [comment]
Any chance this could go on to npm? 

Quite challenging to manually download and build and then import in a docker environment (and manage the dependencies with for an example webpack).

**Comment by tolexy** [comment]
Closely following the CSP-friendly version of this. Is there a general timeline we can expect to easily utilize this much anticipated CSP version?

**Comment by andreseduardop** [comment]
The CSP version will take Alpinejs to the next level. :sparkle:

**Comment by tomb1n0** [comment]
Any news on if the CSP build will be published to NPM soon? I could really do with this on a project 😅

**Comment by lursyy** [comment]
I am building a chrome extension using the current manifest v3, which [disallows unsafe-eval for extension pages](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#:~:text=MV3%20disallows%20certain%20CSP%20modifications). Without this, I am forced to abandon AlpineJS for now :(

**Comment by HugoDF** [comment]
> I am building a chrome extension using the current manifest v3, which [disallows unsafe-eval for extension pages](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#:~:text=MV3%20disallows%20certain%20CSP%20modifications). Without this, I am forced to abandon AlpineJS for now :(

Alpine v3 should have a CSP friendly build where you define your JS in JS so there is no eval.

**Comment by mansoorkhan96** [comment]
> > I am building a chrome extension using the current manifest v3, which [disallows unsafe-eval for extension pages](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#:~:text=MV3%20disallows%20certain%20CSP%20modifications). Without this, I am forced to abandon AlpineJS for now :(
> 
> Alpine v3 should have a CSP friendly build where you define your JS in JS so there is no eval.

How can we install the CSP build? `npm install @alpinejs/csp` does not seem to be working.

**Comment by ryangjchandler** [comment]
@mansoorkhan96 The CSP build hasn't been officially released & published yet.

**Comment by mansoorkhan96** [comment]
> 

Thanks @ryangjchandler for the prompt response. Any workaround? I need to use alpine in a chrome extension and the v3 manifest does not allow `"unsafe-eval"`.

**Comment by ryangjchandler** [comment]
@mansoorkhan96 You could potentially use a submodule of the Alpine repo and build the CSP version from source?

**Comment by mansoorkhan96** [comment]
> @mansoorkhan96 You could potentially use a submodule of the Alpine repo and build the CSP version from source?

I am not sure how to do that. But i am willing to learn, any resource/article where i can get started and create a submodule.

**Comment by ryangjchandler** [comment]
The git documentation is the best place to start: https://git-scm.com/book/en/v2/Git-Tools-Submodules

**Comment by mansoorkhan96** [comment]
> The git documentation is the best place to start: https://git-scm.com/book/en/v2/Git-Tools-Submodules

Thanks @ryangjchandler I will try it.

**Comment by thibaudcolas** [comment]
I got really confused by the unavailability of the CSP build as well, so have proposed a PR to the documentation so it’s at least noted there this is still in progress: #2619.

**Comment by JPGarCar** [comment]
Any update on the availability of the CSP build? I am also unable to use this amazing library due to my strict CSP policy.

**Comment by mittelgrau** [comment]
Any recent development regarding the CSP-build?

**Comment by stepanjakl** [comment]
https://alpinejs.dev/advanced/csp, but I don't think it's released yet 🤷‍♂️

**Comment by curtismorte** [comment]
@stepanjakl looks like you can find the source code here:
https://github.com/alpinejs/alpine/tree/main/packages/csp

I was just doing some testing with the CSP package and the results weren't positive. 

When using `x-data` for reusable components, you can't pass initial parameters. Per the example from the CSP page:
```html
<!-- Good -->
<div x-data="counter">
    <button @click="increment">Increment</button>
 
    <span x-text="count"></span>
</div>

<!-- Won't Work -->
<div x-data="counter(10)">
    <button @click="increment">Increment</button>
 
    <span x-text="count"></span>
</div>
```

```javascript
// Works
Alpine.data('counter', () => ({
    count: 1,
 
    increment() { this.count++ }
}))

// Doesn't work
Alpine.data('counter', (initialValue) => ({
    count: initialValue,
 
    increment() { this.count++ }
}))
```

Thought I would share this. The issue here is passing values that can modify the data object which is used for the 'counter' value. Without looking into the code, my assumption is that the value `counter(10)` doesn't match with the key for data, thus, nothing happens. You could potentially do something like this however:

```html
<!-- Good -->
<div x-data="counter" x-data-arguments='[10]'>
    <button @click="increment">Increment</button>

    <span x-text="count"></span>
</div>
```

```javascript
Alpine.data('counter', () => ({
    count: 0,

    init() {
        let args = this.$el.getAttribute('x-data-arguments');
        if(args){
            try{
                args = JSON.parse(args);
                this.count = args[0];   
            } catch (e) {
                console.error('There was an issue parsing arguments for "counter". ', e);
            }
        }
    },
    
    increment() { this.count++ }
}))
```
> The issue with this solution is that there is still potential to modify values prior to the initialization method happening. Albeit small, still something to point out. I don't see how this is any different than vanilla JS initializing something that gets its values via data attributes during initialization.

Has anyone else actually put the CSP to the test against legitimate use cases like I mentioned?
> Yes I understand this requires you to build the repo from source.

------

Aside, you can build the repo from source by doing the following:
```
# Download source bundle (so many different ways to do this)
gh repo clone alpinejs/alpine

# Install packages
cd alpine && npm install

# Build Alpine
npm run build
``` 

Once you have done this, you can import Alpine like so (assuming you match your path):
```
// Alpine JS
import Alpine from '../../alpine/packages/csp';

// Make Alpine available to the rest of the components
window.Alpine = Alpine;

Alpine.data('counter', () => ({
    count: 0,

    init() {
        let args = this.$el.getAttribute('x-data-arguments');
        if(args){
            try{
                args = JSON.parse(args);
                this.count = args[0];
            } catch (e) {
                console.error('There was an issue parsing arguments for "counter". ', e);
            }
        }
    },

    increment() { this.count++ }
}));

/* -----------------------------------------------------------
 * Initialize Alpine after you have explicitly defined data
 * properties for all module includes.
 * ----------------------------------------------------------- */
Alpine.start();
```

**Comment by gitRup-ca** [comment]
Yes ... I am glad I found this thread after two days of research.  Looks like I am sticking to more study time learning vanilla JS and building my own drop downs and modals etc.   If Alpine.js is not an out-of-box strict CSP compliant solution ...no thanx.  And no thanx to LiveWire too.  Why would I want to learn advanced Alpine/LiveWire over a better understanding of vanilla JS ?

However, I did not come here just to complain.

May I suggest, instead of trying to fix ALL of Alpine.js for a strict CSP,  just include some PREBUILT strict CSP compliant common  needs.

For example;
My client wanted a "hamburger" ( you know the three lines that drop a menu down) for logins etc on your cell phone.  What I ended up doing to make it strict CSP compliant; is using JS onload to see if there is any html element with an ID of MainNavBar.  If there was - then find out if there were any DIV ID's that matched NavMenu1 thru NavMeun9  on DIVs.  This of course all being in an EXTERNAL JS file AND Nonce'd for the hell of it (no inline scripts).  Done with Laravel Spatie - CSP, Tailwind  and Blade components.

Now,  if all I had to do was use STRICT Alpine.js/Blade SYNTAX and I got strict CSP...... SWEEEEEET HEAVEN !

Lets, consider the random website visitor.  They come and get some nice tabs, nice carousal, tool tips, whatever is currently strict CSP possible.   However, now they create an account and login.  Some CSP gets relaxed.   But now they opted for 2FA and THEY choose to relax the CSP opting for more interactivity.  As the User becomes more VERIFIED I would like to give them more Alpine.js features.  And now they are a verified 2FA with a verified credit card. 

Just give me SOME strict CSP ...Alpine stuff AND let me choose when I will relax stuff.  STOP trying to fix it all at once.

x-safe-hamburger-5menu-items
requires DIV ID Parent blade component to be named EXACTLY as above
there MUST BE EXACTLY - 5 child DIV's called EXACTLY x-div-item1 to EXACTLY x-div-item-5.

No less - no more.

Just give me some KEWL shit to make sales.   STOP trying to fix it all at once.

x-safe-modal
display blade msg
John B.
